deepspeed/runtime/engine.py

```diff
@@ -2786,7 +2789,7 @@ def get_sparse_tensor_module_names(original_set, loaded_set, original_parameters
 
         client_state = {key: value for key, value in checkpoint.items() if not key in deepspeed_states}
 
-        if not load_optimizer_states and not load_module_only:
+        if optim_checkpoint is not None:
             client_state['optimizer'] = optim_checkpoint['optimizer']
```

Comment:
The original idea of `client_state` was to provide a convenient way for the checkpoint to contain both deepspeed-specific state, and client-specific state. Rather than the client creating a separate checkpoint file for its state, we designed `deepspeed.save_checkpoint(...)` to accept a `dict` containing client-specific state, so that deepspeed could checkpoint such state on behalf of the client, alongside the deepspeed-specific state. At load time, we identify `client_state` using this hardcoded [logic](https://github.com/microsoft/DeepSpeed/blob/7711bdbbd27c62ab4986f35c1ed01a0268fed92f/deepspeed/runtime/engine.py#L2795).  Unfortunately, at some point this [logic](https://github.com/microsoft/DeepSpeed/blob/7711bdbbd27c62ab4986f35c1ed01a0268fed92f/deepspeed/runtime/engine.py#L2797-L2798) was added to further modify the `client_state`. I am suspicious of this change since `optimizer` is almost always `deepspeed_state` and not `client_state. Nevertheless, can you please revert this test condition for backwards-compatibility concerns. I don't think it should affect the correctness of your changes, since `optim_checkpoint=None` will only be True if `load_module_only == True`. Am I correct?

Reply:
I see, reverting my change then. My understanding was that `client_state` is something we return to the user to indicate what state has been used for which part in the end.

Reply:
> I see, reverting my change then. My understanding was that `client_state` is something we return to the user to indicate what state has been used for which part in the end.

Thanks! Apologies for this confusing feature. The following docs might explain a bit more: 

https://deepspeed.readthedocs.io/en/latest/model-checkpointing.html#deepspeed.DeepSpeedEngine.save_checkpoint
https://deepspeed.readthedocs.io/en/latest/model-checkpointing.html#deepspeed.DeepSpeedEngine.load_checkpoint

Reply:
Unfortunately, reverting this bit breaks several tests, @tjruwase and here's why I think. For `load_module_only=False` and `load_optimizer_states=False` it will attempt to access `optim_checkpoint`, but I'm not quite sure why. Didn't we specify not to use optimizer states - to refresh it from model weights? The intention of this change is to avoid loading the optimizer states since they are not used and hence don't return it via `client_state`.

Reply:
There are other inconsistencies by the way. I thought about including the changes for them as well, but decided to do it in a separate PR. E.g. `deepspeed_states` adds `data_sampler` (key) unconditionally, but it's only loaded into state dict if present in `checkpoint`.

Reply:
Please undo the revert. Your original change is at least meaningful compared to the old code. The entire function needs a rewrite :(.